File: Medium/92.反转链表-ii.cpp
/*
 * @lc app=leetcode.cn id=92 lang=cpp
 *
 * [92] 反转链表 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    ListNode *reverseBetween(ListNode *head, int m, int n)
    {
        //特殊情况
        if (head == NULL)
            return head;
        ListNode *pre = NULL;
        ListNode *cur = head;
        //找到m所在的位置
        while (m > 1)
        {
            pre = cur;
            cur = cur->next;
            m--;
            //n也要同时减一
            n--;
        }
        //这两个指针是用来标记需要反转的那一小段链表的头和尾的
        //这里采用的方法是全部进行反转，到最后再对这一小段的头和尾进行特殊处理
        ListNode *prev = pre;
        ListNode *tail = cur;
        //用于标记cur的下一个节点，防止断链
        ListNode *r = NULL;
        while (n > 0)
        {
            //防止断链
            r = cur->next;
            //进行反转
            cur->next = pre;
            pre = cur;
            cur = r;
            n--;
        }
        //对需要反转的那一小段链表的头和尾节点进行特殊的处理
        if (prev != NULL)
            prev->next = pre;
        else
            head = pre;
        tail->next = cur;
        return head;
    }
};
// @lc code=end

File: Medium/92.反转链表-ii.go
/*
 * @lc app=leetcode.cn id=92 lang=golang
 *
 * [92] 反转链表 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
// 1.pre始终指向第一个待反转节点的前置节点
// 2.cur始终指向第一个待反转节点
// 3.next指向cur节点的下一个节点
//  扫描的同时进行反转，分三步：
// 1.cur的下一个节点指向next的下一个节点
// 2.把next的下一个节点指向pre的下一个节点
// 3.把pre的下一个节点指向next

func reverseBetween(head *ListNode, left int, right int) *ListNode {
	// 用一个虚拟的头节点统一操作
	dummyNode := &ListNode{Val: -1}
	dummyNode.Next = head
	pre := dummyNode
	// 找到第一个待反转的节点，pre指向其前置节点
	for i := 0; i < left-1; i++ {
		pre = pre.Next
	}
	cur := pre.Next
	//  扫描的同时进行反转，分三步：
	// 1.cur的下一个节点指向next的下一个节点
	// 2.把next的下一个节点指向pre的下一个节点
	// 3.把pre的下一个节点指向next
	for i := 0; i < right-left; i++ {
		next := cur.Next
		cur.Next = next.Next
		next.Next = pre.Next
		pre.Next = next
	}
	return dummyNode.Next
}

// @lc code=end

